File: src/gui/pane/CellPane.java
package gui.pane;

import java.util.List;

import javafx.scene.Node;
import javafx.scene.control.ScrollPane;
import javafx.scene.layout.Pane;

import static java.lang.Integer.max;
import static java.lang.Math.ceil;
import static java.lang.Math.floor;
import static javafx.scene.control.ScrollPane.ScrollBarPolicy.AS_NEEDED;
import static javafx.scene.control.ScrollPane.ScrollBarPolicy.NEVER;
import static util.async.Async.runLater;
import static util.functional.Util.forEachWithI;
import static util.graphics.Util.setAnchors;

/**
 * Pane displaying a grid of cells - nodes of same size. Similar to
 * {@link javafx.scene.layout.TilePane}. The cells are always of specified size,
 * vertical gap as well, but horizontal gap is adjusted so the cells are laid
 * out in the whole horizontal space.
 */
public class CellPane extends Pane {
	double cellW = 100;
	double cellH = 100;
	double cellG = 5;

	/**
	 * @param cellWidth cell width
	 * @param cellHeight cell height
	 * @param cellGao cell gap. Vertically it will be altered as needed to maintain
	 * layout.
	 */
	public CellPane(double cellWidth, double cellHeight, double cellGao) {
		cellW = cellWidth;
		cellH = cellHeight;
		cellG = cellGao;
	}

	@Override
	protected void layoutChildren() {
		double width = getWidth();
		List<Node> cells = getChildren();

		int elements = cells.size();
		if (elements==0) return;

		int c = (int) floor((width+cellG)/(cellW+cellG));
		int columns = max(1,c);
		double gapX = cellG +(width+cellG -columns*(cellW+cellG))/columns;
		double gapY = cellG;

		forEachWithI(cells, (i,n) -> {
			double x = i%columns * (cellW+gapX);
			double y = i/columns * (cellH+gapY);
			n.relocate(x,y);
			n.resize(cellW, cellH);
		});

		int rows = (int) ceil(elements/(double)columns);

		runLater(()->setPrefHeight(rows*(cellH+gapY)));
	}

	/** Puts this pane to scrollbar and returns it. */
	public ScrollPane scrollable() {
		ScrollPane s = new ScrollPane();
		s.setContent(this);
		s.setFitToWidth(true);
		s.setFitToHeight(false);
		s.setHbarPolicy(NEVER);
		s.setVbarPolicy(AS_NEEDED);
		getChildren().add(s);
		setAnchors(s,0d);
		return s;
	}

}

File: src/gui/pane/CellPane.kt
package gui.pane

import javafx.scene.control.ScrollPane
import javafx.scene.control.ScrollPane.ScrollBarPolicy.AS_NEEDED
import javafx.scene.control.ScrollPane.ScrollBarPolicy.NEVER
import javafx.scene.layout.Pane
import util.async.runLater
import util.graphics.setAnchors
import java.lang.Integer.max
import java.lang.Math.ceil
import java.lang.Math.floor

/**
 * Pane displaying a grid of cells of the same size. Similar to [javafx.scene.layout.TilePane], the cells are always
 * of specified size, vertical gap too, but horizontal gap is adjusted so the cells are evenly laid out.
 */
class CellPane: Pane {

    val cellW: Double
    val cellH: Double
    val cellG: Double

    /**
     * @param cellWidth cell width
     * @param cellHeight cell height
     * @param cellGap cell gap. Vertically it will be altered as needed to maintain layout.
     */
    constructor(cellWidth: Double, cellHeight: Double, cellGap: Double): super() {
        cellW = cellWidth
        cellH = cellHeight
        cellG = cellGap
    }

    override fun layoutChildren() {
        val elements = children.size
        if (elements==0) return
        val w = width
        val c = floor((w+cellG)/(cellW+cellG)).toInt()
        val columns = max(1, c)
        val gapX = cellG+(w+cellG-columns*(cellW+cellG))/columns
        val gapY = cellG

        children.withIndex().forEach { (i, n) ->
            val x = i%columns*(cellW+gapX)
            val y = i/columns*(cellH+gapY)
            n.relocate(x, y)
            n.resize(cellW, cellH)
        }

        val rows = ceil(elements.toDouble()/columns).toInt()

        runLater { prefHeight = rows*(cellH+gapY) }
    }

    /**
     * Wrap the content in a scroll pane
     *
     * @return scroll pane wrapper of the content
     */
    fun scrollable(): ScrollPane {
        val s = ScrollPane().apply {
            content = this
            isFitToWidth = true
            isFitToHeight = false
            hbarPolicy = NEVER
            vbarPolicy = AS_NEEDED
        }
        children += s
        s.setAnchors(0.0)
        return s
    }

}